File: C/sorting/quicksort.c
#include "quicksort.h"

static int partition(int *a, int l, int r);
static void quicksort_rec(int *a, int l, int r);
static void swap(int *i, int *j);

static int partition(int *a, int l, int r) {
    int i = l;
    int j;
    for (j = l; j < r; j++) {
        if (a[j] <= a[r]) {
            swap(&a[i], &a[j]);
            i++;
        }
    }
    swap(&a[i], &a[r]);
    return i;
}

static void quicksort_rec(int *a, int l, int r) {
    if (l < r) {
        int m = partition(a, l, r);
        quicksort_rec(a, l, m - 1);
        quicksort_rec(a, m + 1, r);
    }
}

static void swap(int *i, int *j) {
    int temp = *i;
    *i = *j;
    *j = temp;
}

void quicksort(int *a, int len) {
    quicksort_rec(a, 0, len - 1);
}

File: Java/ZahlensortiererGraphisch/src/Quicksort.java
import javax.swing.JOptionPane;

public final class Quicksort extends AbstractSorter {

    private int count = 0;

    public void run() {
        count = 0;
        quicksort(0, array.length - 1);
        JOptionPane.showMessageDialog(
            intArrayGUI,
            "Sortierung beendet! (Vertauschungen " + count + ")"
        );
        intArrayGUI.setSorted(true);
    }

    /**
     * Eine Hilfsmethode fuer den QuickSort-Algorithmus. Sie sortiert das Array
     * in der Weise vor, dass alle Elemente kleiner dem Pivot-Element auf die
     * linke Seite und alle groessen auf die rechte Seite desselben verschoben
     * werden.
     *
     * @param start Hier beginnt die Sortierung.
     * @param end   Hier endet die Sortierung.
     * @return Den neuen Index des Pivot-Elements.
     */
    private int divide(final int start, final int end) {
        final int pivot = array[end];
        int i = start;
        int j = end - 1;
        while (i < j) {
            for ( ; i < end && array[i] <= pivot; i++);
            for ( ; j > start && array[j] >= pivot; j--);
            if (i < j) {
                swap(i, j);
                count++;
            }
        }
        if (array[i] > pivot) {
            swap(i, end);
            count++;
        }
        return i;
    }

    /**
     * Der Sortieralgorithmus "QuickSort". Die Methode ruft sich rekursiv
     * erneut mit veraenderten Grenzen auf, um das Array nach und nach zu
     * sortieren.
     *
     * @param start
     * @param end
     */
    private void quicksort(final int start, final int end) {
        if (start < end) {
            final int pos = divide(start, end);
            quicksort(start, pos - 1);
            quicksort(pos + 1, end);
        }
    }
}
